### PR TITLE
Rollup fixes to specification, Update queries, bloom filters, negative proof semantics

### PR DESCRIPTION
**Pull request by fehlmach**

### CODE REVIEW

File: draft-trammell-rains-protocol.md [hunk 1]
```diff
@@ -392,30 +413,66 @@ For efficiency's sake, information elements within a shard common to all
 assertions (zone, context, signature) within the shard may be omitted from the
 assertions themselves.
 
+A bloom filter is a space-efficient probabilistic data structure represented as
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: We should call this something else in the spec: "bloom filter" is indeed how it is implemented, but what it actually provides is a "probabalistic negative proof".

File: draft-trammell-rains-protocol.md [hunk 2]
```diff
@@ -392,30 +413,66 @@ For efficiency's sake, information elements within a shard common to all
...
+For efficiency's sake, information elements within a zone common to all
+assertions, shards and bloom filters (zone, context, signature) within the zone
+may be omitted from the assertions, shards and bloom filters themselves.
+
+A bloom filter can only be used as a negative answer. A shard should only be
+used as a proof of nonexistence in a query response. A zone can be used for both
+positive and negative answers as the contained information conceptually belongs
+together.
+
+### Semantic of nonexistence proofs
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: think about moving this up?

File: draft-trammell-rains-protocol.md [hunk 3]
```diff
@@ -365,12 +384,12 @@ A signature over an assertion contains the following information elements:
   the algorithm used.
 
 The signature protects all the information in an assertion as well as its own
-algorithm identifier, keyspace, keyphase, valid-since, and valid-until values;
-it does not protect other signatures on the assertion.
+algorithm identifier, keyspace identifier, key phase, valid-since, and
+valid-until values; it does not protect other signatures on the assertion.
 
-### Shards and Zones
+### Shards, Bloom Filters, and Zones {#shards-bloom-filters-zones}
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: These should be two separate sections:

1. shards and whatever-we'll-call-bloomfilters (useful for negative proof only)
2. zones (complete grouping, useful for pos + neg, common in the tiny-zone case)

File: draft-trammell-rains-protocol.md [hunk 4]
```diff
@@ -593,24 +716,31 @@ and notification maps is given in the symbol table below:
...
-| 8    | query-name     | Fully qualified name for a query              |
-| 10   | query-types    | Acceptable object types for query             |
-| 11   | shard-range    | Lexical range of Assertions in Shard          |
-| 12   | query-expires  | Absolute timestamp for query expiration       |
-| 13   | query-opts     | Set of query options requested                |
-| 21   | note-type      | Notification type                             |
-| 22   | note-data      | Additional notification data                  |
-| 23   | content        | Content of a message, shard, or zone          |
+| Code | Name           | Description                                          |
+|-----:|----------------|------------------------------------------------      |
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: consider whether any of these should be >23

File: draft-trammell-rains-protocol.md [hunk 5]
```diff
@@ -745,23 +881,120 @@ and MUST end with '.' (the root zone). If not present, the zone of the assertion
...
+function in the optimization and the second part to the second one. Thus, the
+following formula is used to obtain the position which will be set to 1
+according to the ith hash function:
+pos = (hash1 + hash2*i) modulo bit-string-size
+
+{: #tabhashfamilies title="Bloom Filter hash families"}
+
+| Code | Name       | Notes                                     |
+|-----:|------------|-------------------------------------------|
+| 1    | fnv-64     | Value contains FNV-64 hash (64 bytes)     |
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: should we not also have SHA for consistency?

File: draft-trammell-rains-protocol.md [hunk 6]
```diff
@@ -826,6 +1065,11 @@ query, as in {{tabqopts}}.
 | 6    | Enable query token tracing                                     |
 | 7    | Disable verification delegation (client protocol only)         |
 | 8    | Suppress proactive caching of future assertions                |
+| 9    | Enforce recursive lookup on assertion cache miss               |
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: rephrase / coalesce 9-13 in terms of goals

File: draft-trammell-rains-protocol.md [hunk 7]
```diff
@@ -848,6 +1092,104 @@ Option 8 specifies that a querier's interest in a query is strictly ephemeral,
...
+The value of the query-opts (13) key, if present, is an array of integers in
+priority order of the querier's preferences in tradeoffs in answering the
+assertion update query, as in {{tabqopts}}. Only query option codes 1, 2, 3, 6,
+7, 8 are allowed.
+
+{: #tabuqhash title="Update Query Hash Function Codes"}
+
+| Code | Name      | Notes                                  |
+|-----:|-----------|----------------------------------------|
+| 1    | sha-256   | Value contains SHA-256 hash (32 bytes) |
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: use the same hash functions everywhere

File: draft-trammell-rains-protocol.md [hunk 8]
```diff
@@ -929,8 +1271,11 @@ contain the token (2) and note-type (21) keys and MAY contain the note-data
 | Code | Description                                                    |
 |-----:|----------------------------------------------------------------|
 | 100  | Connection heartbeat                                           |
+| 200  | The hashed section in an update query is still fine            |
+| 210  | The hashed assertion has been revoked and is no longer valid   |
+| 211  | Section proving nonexistence is being updated                  |
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: generalize this -> "more specific information may follow"

File: draft-trammell-rains-protocol.md [hunk 9]
```diff
@@ -1405,48 +1806,64 @@ If the server or client can parse the message, it:
...
-  the reply for the forwarded query.
+1. determines whether it has expired by checking the query-expires value. If so,
+   it drops the query silently. If not, it:
+2. determines whether it has at least one stored assertion answering the query.
+   If so, it returns the assertion(s) with the longest validUntil value that is
+   already valid. If not, it:
+3. checks whether the query specifies option 9-12. If so, it:
+   - determines whether the chosen option is in compliance with the server's
+     configuration and policy. If so, and:
+     - option 9 is set it continues with step 4.
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: rework per options above

File: draft-trammell-rains-protocol.md (hunk 1 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: It is called now pshard (probabilistic shard)

File: draft-trammell-rains-protocol.md (hunk 4 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: Why did you chose this range? "Integer map keys in the range -22 to +23
are reserved for the use of future versions or extensions to the RAINS protocol"
We have about 25 entries left. Is that enough?

**Pull request merged**